### PR TITLE
Capture all Error console.log events

### PR DESCRIPTION
**Pull request by PeterYinusa**
## Explanation

This PR surfaces all the errors found in the browser

Some errors were previously missed (e.g. `Unable to find value of key "welcomeBack" for locale "en"`)

In this PR you can see some new errors are being caught that weren't previously. [Here](https://app.circleci.com/pipelines/github/MetaMask/metamask-extension/36634/workflows/8ee91db2-b761-45d7-858d-96ade2eb8d8a/jobs/984094) and [here](https://app.circleci.com/pipelines/github/MetaMask/metamask-extension/36634/workflows/8ee91db2-b761-45d7-858d-96ade2eb8d8a/jobs/984063).
With the corresponding fix to ignore the error [here](https://github.com/MetaMask/metamask-extension/pull/17710/commits/29fd02211a30a2293096ddeaed80afcd896dd469) and [here](https://github.com/MetaMask/metamask-extension/pull/17710/commits/8a49cfaeb4b2aa92e37b9855b9c0d7e78217056f)

The following enhancements have been made:
- Use the Chrome DevTools Protocol to listen to console events, rather than searching the browser logs
- Throw the error immediately once found

**Drawback:**
The `--leave-running` flag may not leave things running `if` the error is thrown **after the test suite has run but before the browser has closed**. 
I believe this is a worthwhile tradeoff since errors that previously weren't surfaced are now shown. This flag is only used for debugging and is not used in CI. The flag will continue to work for the majority of cases

### Before

```
peteryinusa@Peters-MacBook-Pro metamask-extension-e2e % yarn test:e2e:single test/e2e/tests/account-details.spec.js --browser chrome
Starting ChromeDriver 109.0.5414.74 (e7c5703604daa9cc128ccf5a5d3e993513758913-refs/branch-heads/5414@{#1172}) on port 56359
Only local connections are allowed.
Please see https://chromedriver.chromium.org/security-considerations for suggestions on keeping ChromeDriver safe.
ChromeDriver was started successfully.

DevTools listening on ws://127.0.0.1:56369/devtools/browser/80fcdeff-ed6a-4064-8e31-4aed10f719e8
objc[5633]: Class WebSwapCGLLayer is implemented in both /System/Library/Frameworks/WebKit.framework/Versions/A/Frameworks/WebCore.framework/Versions/A/Frameworks/libANGLE-shared.dylib (0x7ffa43f763b0) and /Applications/Google Chrome.app/Contents/Frameworks/Google Chrome Framework.framework/Versions/110.0.5481.96/Libraries/libGLESv2.dylib (0x1115fc220). One of the two will be used. Which one is undefined.
[1676389156.904][WARNING]: This version of ChromeDriver has not been tested with Chrome version 110.
<testsuite name="Mocha Tests" tests="1" failures="0" errors="1" skipped="0" timestamp="Tue, 14 Feb 2023 15:39:30 GMT" time="15.405">
<testcase classname="Show account details" name="should show the QR code for the account" time="15.39"><failure>Waiting for element to be located By(css selector, #password)
Wait timed out after 10027ms
TimeoutError: Waiting for element to be located By(css selector, #password)
Wait timed out after 10027ms
    at /Users/peteryinusa/Documents/GitHub/metamask-extension-e2e/node_modules/selenium-webdriver/lib/webdriver.js:906:17
    at processTicksAndRejections (node:internal/process/task_queues:96:5)</failure></testcase>
</testsuite>
Error: Exited with code '1'
    at runInShell (/Users/peteryinusa/Documents/GitHub/metamask-extension-e2e/development/lib/run-command.js:128:29)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async /Users/peteryinusa/Documents/GitHub/metamask-extension-e2e/test/e2e/run-e2e-test.js:97:5
    at async retry (/Users/peteryinusa/Documents/GitHub/metamask-extension-e2e/development/lib/retry.js:30:7)
    at async main (/Users/peteryinusa/Documents/GitHub/metamask-extension-e2e/test/e2e/run-e2e-test.js:96:3)
Error: Retry limit reached
    at retry (/Users/peteryinusa/Documents/GitHub/metamask-extension-e2e/development/lib/retry.js:39:9)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async main (/Users/peteryinusa/Documents/GitHub/metamask-extension-e2e/test/e2e/run-e2e-test.js:96:3)
```
### After

```
peteryinusa@Peters-MacBook-Pro metamask-extension-e2e % yarn test:e2e:single test/e2e/tests/account-details.spec.js --browser chrome
Starting ChromeDriver 109.0.5414.74 (e7c5703604daa9cc128ccf5a5d3e993513758913-refs/branch-heads/5414@{#1172}) on port 56132
Only local connections are allowed.
Please see https://chromedriver.chromium.org/security-considerations for suggestions on keeping ChromeDriver safe.
ChromeDriver was started successfully.

DevTools listening on ws://127.0.0.1:56147/devtools/browser/a2e88b49-0a50-46c7-811e-cca1dee38660
objc[4921]: Class WebSwapCGLLayer is implemented in both /System/Library/Frameworks/WebKit.framework/Versions/A/Frameworks/WebCore.framework/Versions/A/Frameworks/libANGLE-shared.dylib (0x7ffa43f763b0) and /Applications/Google Chrome.app/Contents/Frameworks/Google Chrome Framework.framework/Versions/110.0.5481.96/Libraries/libGLESv2.dylib (0x10b93c220). One of the two will be used. Which one is undefined.
[1676388587.437][WARNING]: This version of ChromeDriver has not been tested with Chrome version 110.
<testsuite name="Mocha Tests" tests="1" failures="0" errors="1" skipped="0" timestamp="Tue, 14 Feb 2023 15:29:51 GMT" time="6.705">
<testcase classname="Show account details" name="should show the QR code for the account" time="6.561"><failure>Error: Unable to find value of key &#x22;welcomeBack&#x22; for locale &#x22;en&#x22;
  at s.getMessage (chrome-extension://eemkccoomdelmfifgfcnkfbjhhfenbpk/ui-7.js:1:286142)
  at chrome-extension://eemkccoomdelmfifgfcnkfbjhhfenbpk/ui-7.js:1:119517
  at E.render (chrome-extension://eemkccoomdelmfifgfcnkfbjhhfenbpk/ui-10.js:1:53229)
  at Qo (chrome-extension://eemkccoomdelmfifgfcnkfbjhhfenbpk/ui-4.js:1:71454)
  at Jo (chrome-extension://eemkccoomdelmfifgfcnkfbjhhfenbpk/ui-4.js:1:71249)
  at Os (chrome-extension://eemkccoomdelmfifgfcnkfbjhhfenbpk/ui-4.js:1:106112)
  at wl (chrome-extension://eemkccoomdelmfifgfcnkfbjhhfenbpk/ui-4.js:1:90773)
  at kl (chrome-extension://eemkccoomdelmfifgfcnkfbjhhfenbpk/ui-4.js:1:90698)
  at dl (chrome-extension://eemkccoomdelmfifgfcnkfbjhhfenbpk/ui-4.js:1:88046)
  at chrome-extension://eemkccoomdelmfifgfcnkfbjhhfenbpk/ui-4.js:1:46504
  at n.unstable_runWithPriority (chrome-extension://eemkccoomdelmfifgfcnkfbjhhfenbpk/ui-4.js:1:467245)
  at Yr (chrome-extension://eemkccoomdelmfifgfcnkfbjhhfenbpk/ui-4.js:1:46213)
  at Jr (chrome-extension://eemkccoomdelmfifgfcnkfbjhhfenbpk/ui-4.js:1:46449)
  at Zr (chrome-extension://eemkccoomdelmfifgfcnkfbjhhfenbpk/ui-4.js:1:46384)
  at hl (chrome-extension://eemkccoomdelmfifgfcnkfbjhhfenbpk/ui-4.js:1:88419)
  at ec (chrome-extension://eemkccoomdelmfifgfcnkfbjhhfenbpk/ui-4.js:1:113602)
  at render (chrome-extension://eemkccoomdelmfifgfcnkfbjhhfenbpk/ui-4.js:1:115519)
  at chrome-extension://eemkccoomdelmfifgfcnkfbjhhfenbpk/ui-7.js:1:385132
Error: Error: Unable to find value of key &#x22;welcomeBack&#x22; for locale &#x22;en&#x22;
  at s.getMessage (chrome-extension://eemkccoomdelmfifgfcnkfbjhhfenbpk/ui-7.js:1:286142)
  at chrome-extension://eemkccoomdelmfifgfcnkfbjhhfenbpk/ui-7.js:1:119517
  at E.render (chrome-extension://eemkccoomdelmfifgfcnkfbjhhfenbpk/ui-10.js:1:53229)
  at Qo (chrome-extension://eemkccoomdelmfifgfcnkfbjhhfenbpk/ui-4.js:1:71454)
  at Jo (chrome-extension://eemkccoomdelmfifgfcnkfbjhhfenbpk/ui-4.js:1:71249)
  at Os (chrome-extension://eemkccoomdelmfifgfcnkfbjhhfenbpk/ui-4.js:1:106112)
  at wl (chrome-extension://eemkccoomdelmfifgfcnkfbjhhfenbpk/ui-4.js:1:90773)
  at kl (chrome-extension://eemkccoomdelmfifgfcnkfbjhhfenbpk/ui-4.js:1:90698)
  at dl (chrome-extension://eemkccoomdelmfifgfcnkfbjhhfenbpk/ui-4.js:1:88046)
  at chrome-extension://eemkccoomdelmfifgfcnkfbjhhfenbpk/ui-4.js:1:46504
  at n.unstable_runWithPriority (chrome-extension://eemkccoomdelmfifgfcnkfbjhhfenbpk/ui-4.js:1:467245)
  at Yr (chrome-extension://eemkccoomdelmfifgfcnkfbjhhfenbpk/ui-4.js:1:46213)
  at Jr (chrome-extension://eemkccoomdelmfifgfcnkfbjhhfenbpk/ui-4.js:1:46449)
  at Zr (chrome-extension://eemkccoomdelmfifgfcnkfbjhhfenbpk/ui-4.js:1:46384)
  at hl (chrome-extension://eemkccoomdelmfifgfcnkfbjhhfenbpk/ui-4.js:1:88419)
  at ec (chrome-extension://eemkccoomdelmfifgfcnkfbjhhfenbpk/ui-4.js:1:113602)
  at render (chrome-extension://eemkccoomdelmfifgfcnkfbjhhfenbpk/ui-4.js:1:115519)
  at chrome-extension://eemkccoomdelmfifgfcnkfbjhhfenbpk/ui-7.js:1:385132
    at driver (test/e2e/webdriver/driver.js:476:9)
    at /Users/peteryinusa/Documents/GitHub/metamask-extension-e2e/test/e2e/webdriver/driver.js:469:16
    at WebSocket.&#x3C;anonymous&#x3E; (node_modules/selenium-webdriver/lib/webdriver.js:1412:9)
    at WebSocket.emit (node:events:539:35)
    at WebSocket.emit (node:domain:475:12)
    at Receiver.receiverOnMessage (node_modules/ws/lib/websocket.js:1178:20)
    at Receiver.emit (node:events:527:28)
    at Receiver.emit (node:domain:475:12)
    at Receiver.dataMessage (node_modules/ws/lib/receiver.js:528:14)
    at /Users/peteryinusa/Documents/GitHub/metamask-extension-e2e/node_modules/ws/lib/receiver.js:479:23
    at /Users/peteryinusa/Documents/GitHub/metamask-extension-e2e/node_modules/ws/lib/permessage-deflate.js:308:9
    at /Users/peteryinusa/Documents/GitHub/metamask-extension-e2e/node_modules/ws/lib/permessage-deflate.js:391:7
    at afterWrite (node:internal/streams/writable:497:5)
    at onwrite (node:internal/streams/writable:477:7)
    at Zlib.cb (node:internal/streams/transform:202:7)
    at Zlib.processCallback (node:zlib:611:8)</failure></testcase>
</testsuite>
Error: Exited with code '1'
    at runInShell (/Users/peteryinusa/Documents/GitHub/metamask-extension-e2e/development/lib/run-command.js:133:29)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async /Users/peteryinusa/Documents/GitHub/metamask-extension-e2e/test/e2e/run-e2e-test.js:100:5
    at async retry (/Users/peteryinusa/Documents/GitHub/metamask-extension-e2e/development/lib/retry.js:30:7)
    at async main (/Users/peteryinusa/Documents/GitHub/metamask-extension-e2e/test/e2e/run-e2e-test.js:99:3)
Error: Retry limit reached
    at retry (/Users/peteryinusa/Documents/GitHub/metamask-extension-e2e/development/lib/retry.js:39:9)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async main (/Users/peteryinusa/Documents/GitHub/metamask-extension-e2e/test/e2e/run-e2e-test.js:99:3)
```
## Manual Testing Steps

1. Check out this branch
2. Remove the [welcomeBack](https://github.com/MetaMask/metamask-extension/blob/e2e-improve-logging/app/_locales/en/messages.json#L4419) translation
3. install dependencies : `yarn`
4. build the extension : `yarn:build:test`
5. run a test, that hasn't turned off failOnConsoleError: `yarn test:e2e:single test/e2e/tests/account-details.spec.js --browser chrome`
6. Verify the error that is thrown in the browser console has surfaced in the terminal console and the xml report `test/test-results/e2e/account-details.spec.js.xml`

## Pre-merge author checklist

- [x] I've clearly explained:
  - [x] What problem this PR is solving
  - [x] How this problem was solved
  - [x] How reviewers can test my changes
- [x] Sufficient automated test coverage has been added

## Pre-merge reviewer checklist

- [ ] Manual testing (e.g. pull and build branch, run in browser, test code being changed)
- [ ] PR is linked to the appropriate GitHub issue
- [ ] **IF** this PR fixes a bug in the release milestone, add this PR to the release milestone

If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.

### CODE REVIEW

File: test/e2e/webdriver/driver.js [hunk 1]
```diff
@@ -481,17 +482,31 @@ class Driver {
...
+          description.includes(message),
+        );
+        if (!ignore) {
+          errors.push(description);
+          logBrowserError(failOnConsoleError, description);
+        }
+      }
+    });
+  }
+}
```
Comment (PeterYinusa) [inline test/e2e/webdriver/driver.js:503]: Continuously listen to the console.log events

File: test/e2e/webdriver/driver.js [hunk 2]
```diff
@@ -481,17 +482,31 @@ class Driver {
...
+
+function logBrowserError(failOnConsoleError, errorMessage) {
```
Comment (PeterYinusa) [inline test/e2e/webdriver/driver.js:505]: Used for both errors and exceptions

File: test/e2e/mv3-perf-stats/init-load-stats.js [hunk 3]
```diff
@@ -22,7 +22,7 @@ async function profilePageLoad() {
   const parsedLogs = {};
   try {
     await withFixtures(
-      { fixtures: new FixtureBuilder().build() },
+      { fixtures: new FixtureBuilder().build(), failOnConsoleError: false },
```
Comment (PeterYinusa) [inline test/e2e/mv3-perf-stats/init-load-stats.js:25]: This test now surfaces console errors

File: test/e2e/webdriver/driver.js [hunk 4]
```diff
@@ -481,17 +482,31 @@ class Driver {
...
-      (entry) => !ignoredLogTypes.includes(entry.level.toString()),
-    );
-    const errorObjects = errorEntries.map((entry) => entry.toJSON());
-    return errorObjects.filter(
-      (entry) =>
-        !ignoredErrorMessages.some((message) =>
-          entry.message.includes(message),
-        ),
-    );
+    const { errors } = this;
```
Comment (mcmire) [inline test/e2e/webdriver/driver.js:485]: This looks like some advanced stuff. Nice!

**Review by mcmire — COMMENTED** [review]
This is a fantastic change. This looks like it'll be an immense help for debugging e2e tests!

**Review by brad-decker — APPROVED** [review]

**Comment by brad-decker** [comment]
this will be super helpful

File: development/lib/run-command.js [hunk 5]
```diff
@@ -99,6 +99,7 @@ async function runInShell(command, args, output) {
       const childProcess = spawn(command, args);
       childProcess.stdout.setEncoding('utf8');
       childProcess.stdout.pipe(process.stdout);
+      childProcess.stderr.pipe(process.stderr);
```
[the remark this answers is not in the document]
Comment (PeterYinusa) [reply development/lib/run-command.js]: This change is already on the develop branch and is no longer shown in this diff

**Review by pedronfigueiredo — APPROVED** [review]

**Pull request merged**